[KOGITO-8818] Rearrange guides to add operator and quarkus cloud (#313)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/operator/build-and-deploy-workflows.adoc

:kogito_greeting_example_url: https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-greeting-quarkus
:kaniko_issue_url: https://github.com/GoogleContainerTools/kaniko/issues/2201

This document describes how to build and deploy your workflow on a Kubernetes cluster using the link:{kogito_serverless_operator_url}[{operator_name}] only by having a workflow definition.

[IMPORTANT]
====
{operator_name} is under active development with features yet to be implemented. Please see xref:cloud/operator/known-issues.adoc[].
====

.Prerequisites
* A workflow definition.
* The {operator_name} installed. See xref:cloud/operator/install-serverless-operator.adoc[]

== Preparing for the build

You should follow these steps to create a container that you can deploy as a service on Kubernetes.

=== Create a namespace for the building phase

Create a new namespace that will hold all the resources that the operator will create (pods, deployments, services, secretes, config map, and Custom Resources) in this guide.

.Create a namespace for the application to build & run in
[source,bash,subs="attributes+"]
----
kubectl create namespace kogito-workflows

> ----

Reviewer: maybe add the tip to set the new namespace to new one?

=== Create a secret for the container registry authentication
.Create a secret for the container registry authentication
[source,bash,subs="attributes+"]
----

> kubectl create secret docker-registry regcred --docker-server=<registry_url> --docker-username=<registry_username> --docker-password=<registry_password> --docker-email=<registry_email> -n kogito-workflows

Reviewer: maybe provide a example with all parameters set, e.g. to quay.io as the registry and dummy credentials?
Or even better, using the  `~/.docker/config.json` file?

Reviewer: oops, the the config file is just below :)

----

or you can directly import your local docker config into your Kubernetes cluster:

.Create a secret for the container registry authentication based on local docker config
[source,bash,subs="attributes+"]
----
kubectl create secret generic regcred --from-file=.dockerconfigjson=$\{HOME\}/.docker/config.json --type=kubernetes.io/dockerconfigjson -n kogito-workflows
----

[WARNING]
====

> Double check your `$\{HOME\}/.docker/config.json`. If you're using local desktop authentication, this configuration **won't work** in the cluster. 

Reviewer: ```suggestion
Double check your `$\{HOME\}/.docker/config.json`. If you're using local desktop authentication, this configuration **won't work** in the cluster. You can initialize this by logging in in the target registry, e.g. `docker login`
```

Reviewer: or maybe podman login.

====

=== Configure the {operator_name} (i.e. registry address, secret) for building your workflows

The `KogitoServerlessPlatform` is the resource used to control the behavior of the {operator_name}.
It defines the behavior of all Custom Resources (Workflow and Build) in the given namespace.

Since the operator is installed in global mode, you will need to specify a platform in each namespace where you want to deploy workflows.

> You can find a basic `KogitoServerlessPlatform` CR example in the `config/samples` folder that you can simply apply to configure your operator.

Reviewer: ```suggestion
You can find a basic `KogitoServerlessPlatform` Custom Resource example in the `config/samples` folder that you can simply apply to configure your operator.
```

Reviewer: add a link to the operator repo?

----
You can check the logs of the build of your workflow via:

.Get the workflow application pod logs
[source,bash,subs="attributes+"]
----
kubectl logs kogito-greeting-builder -n kogito-workflows
----

The final pushed image must be printed into the logs at the end of the build.

== Check the workflow application is running

In order to check that the {product_name} Greeting application is up and running, you can try to perform a test HTTP call, from the greeting pod.

.Check the greeting application is running
[source,bash,subs="attributes+"]
----
kubectl patch svc greeting -n kogito-workflows -p '{"spec": {"type": "NodePort"}}'
GREETING_SVC=$(minikube service greeting -n kogito-workflows --url)
curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d '{"name": "John", "language": "English"}' $GREETING_SVC/greeting
----

If everything is working well you should receive a response like this:

.Response from the greeting application
[source,json,subs="attributes+"]
----
{"id":"b5fbfaa3-b125-4e6c-9311-fe5a3577efdd","workflowdata":{"name":"John","language":"English","greeting":"Hello from JSON Workflow, "}}
----

> // TODO: add a troubleshooting guide - https://issues.redhat.com/browse/KOGITO-8864

Reviewer 2: isn't that done ?

Reviewer 3: @radtriste JIRA at least still open, @davidesalerno could you confirm this is still pending?

Reviewer 4: yes

> * xref:cloud/operator/known-issues.adoc[]
> * xref:cloud/operator/developing-workflows.adoc[]

Reviewer 2: ```suggestion
== Additional resources

* xref:cloud/operator/known-issues.adoc[]
* xref:cloud/operator/developing-workflows.adoc[]
```
